### PR TITLE
Sort available jobs

### PR DESCRIPTION
**Pull request by alextd**
## Summary
New UI to sort the available jobs list.

## Save File
(Literally any save would see this update)

## Artwork Checklist
 - [x] ~I updated the copyright attributions~, or **decline to claim copyright of any assets produced or modified** (just some basic UI I whipped up for this)
 - [x] I created a PR to the [endless-sky-assets repo](https://github.com/EndlessSkyCommunity/endless-sky-assets) with the necessary image, blend, and texture assets: {{[PR link](https://github.com/EndlessSkyCommunity/endless-sky-assets/pull/44)}}
 - [x] I created a PR to the [endless-sky-high-dpi repo](https://github.com/endless-sky/endless-sky-high-dpi) with the `@2x` versions of these art assets: {{[PR link](https://github.com/endless-sky/endless-sky-high-dpi/pull/256)}}

-----------------------
**Feature:** This PR implements the feature request ~detailed and discussed in issue #{{insert number}}~ by me, right now.

## Feature Details
Sort available jobs by name, pay, speed (ie. shortest path), or "convenience" (if the youre already going to the destination)

Also, checkboxes to separate jobs to the bottom of the list: jobs you can't accept (e.g. not enough cargo space) and rush jobs with a deadline (for when you're too busy)

Tag any rush jobs in the list with 'fast forward' arrow.

Sort Settings are saved to the pilot file.

Some code details: 
- Saved the calculated expected # of jumps, as we can now sort by that value. It's not saved to file but recalculated on load.
- Exposed MissionAction's payment by just referring to its GameAction's Payment()
- Added "apparent payment" attribute to missions - e.g. for the wolf/cabbage mission, so the sorted payment value matches what is shown in the description, which can be different from the actual mission payment

## UI Screenshots
**Before**
image

**After**
Here, the missions are sorted by greatest pay, and the greyed-out 'can't accept' jobs are separated to the bottom of the list.
image
Hopefully the UI makes sense and isn't too cluttered. It could easily be its own row. But I like it.
The new UI elements there are: 
- A checkbox to include 'rush' jobs with a deadline
- A checkbox to include greyed-out 'can't accept' jobs
- A button to select sort type ($ = sort by Pay)
- An up/down button to toggle sort direction.
- Tags on any rush jobs.

Here the sorting is done by speed ( fewest jumps from Betelgeuse ). Note that the 'Marauders' job has no determinable path, therefore is sorted to the end of the list. Also, The checkbox for rush jobs is off, so they are separated at the bottom as well.
image

This is how the UI would default to: including both checkboxes and sorting by name.
image

Selected to ignore unacceptable jobs and sort by pay
image

## Usage Examples
You want to take the biggest pay mission: Sort by pay.
You want to take a quick mission: Sort by speed.
You are too busy to accept rush missions: Checkbox to separate rush jobs.
You don't want to be bothered by missions you don't have room for: Checkbox to separate those jobs.
You only want to finish up your current jobs, accepting new missions only if you're already headed to their destination: Sort by 'convenient'

## Testing Done
Lots - found many issues that were fixed.
Completed missions, accepted new missions, completed again.
Clicked around and turned settings on/off . Sorting was inconsistent, so I made it consistent.
Saving/Loading saves the chosen sort method to file. Needed to recalculate # of jumps after galaxy was updated instead of while the jobs were loaded, or data would be wrong (e.g. a mission to Deneb was thought unreachable )
Jobs that go through wormholes have a negative 'jumps' count - those are considered to be slower and sort to the end of the list.

## Performance Impact
N/A - It's only sorted when clicked, and it's not exactly a large set of things to sort.

### CODE REVIEW

**Comment by alextd** [comment]
So I wonder if the UI is clear enough. I think after clicking the checkboxes on and off it's easy to see what those do. Sorting by name or pay is obvious too. The third sorting is by 'speed', as in, fewest jumps. That icon I whipped up hopefully looks like a speedy planet. And there are odd cases that don't sort well - bounty missions have an unknown number of jumps, so are at the end of the list, which would be even past. Missions through wormholes are also considered even slower.

I am considering tooltips for the new UI elements, or moving them to a second row with more text.
Also I did not make 2x sized assets so that's also something to do.

**Comment by alextd** [comment]
I just remembered a possible issue - sorting by distance when you haven’t mapped the galaxy. It’s not much, but it could show that certain systems are not as close as they seem. Well I am not considering it not a problem.

**Comment by MasterOfGrey** [comment]
`Sort Settings are saved to the pilot file.` Perfection

**Comment by alextd** [comment]
Also. If the mission text doesn't mention pay - it still sorts by pay. A bit of a head-scratcher what that's supposed to mean.

**Comment by alextd** [comment]
image

New "convenient" sorting, showing you missions where you're already going. It prioritizes missions with a destination planet that matches one of your current missions.

**Comment by MasterOfGrey** [comment]
> New "convenient" sorting, showing you missions where you're already going. It prioritizes missions with a destination planet that matches one of your current missions.

Now **_this_** is the QoL update we need. =D

**Comment by alextd** [comment]
I am considering an idea that adds an extra chunk to the game : expandable UI

So, what's in the game now would be the starting point, and you could earn UI upgrades like this Mission Sorter, probably in some mission chain once you have enough cargo space in your fleet. This idea comes with my other project I'm working on: auto-buying commodities. That would add a button to the map screen to buy the best commodity which nets the most profit when sold at the selected planet. And possibly an automated version of that.

I wonder what other silly widgets could be an option. What sort of thing do I want to see, but shouldn't be the default, not something for new players or, lore-wise, something inexperienced rookie pilots would use. But a captain of a cargo fleet who has got connections and the latest gadgets would have this sort of stuff, which is represented with new UI.

Anywho it's late and I'm throwing this zany idea out there

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -1559,6 +1567,66 @@ const list<Mission> &PlayerInfo::AvailableJobs() const
...
+
+void PlayerInfo::ToggleAvailableSortAsc()
+{
+	availableSortAsc = !availableSortAsc;
+	SortAvailable();
+}
+
+
+
+const bool PlayerInfo::SeparateRush() const
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:1600]: Shouldn't these getters should start with `Get`?

File: source/MissionPanel.cpp [hunk 2]
```diff
@@ -491,7 +517,9 @@ bool MissionPanel::Hover(int x, int y)
 		if(static_cast<int>(index) < AcceptedVisible())
 			dragSide = 1;
 	}
-	return dragSide ? true : MapPanel::Hover(x, y);
+	hoverSort = y < Screen::Top() + 30 && y >= Screen::Top() + 10 &&
+			x < Screen::Left() + SIDE_WIDTH - 25 && x >= Screen::Left() + SIDE_WIDTH - 45;
+	return (dragSide || hoverSort) ? true : MapPanel::Hover(x, y);
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:522]: Is this change required? I am not sure, but I think that if it is selected we will fall into the first if and return false/true already

File: source/Mission.cpp [hunk 3]
```diff
@@ -1189,7 +1196,7 @@ Mission Mission::Instantiate(const PlayerInfo &player, const shared_ptr<Ship> &b
 	{
 		const Trade::Commodity *commodity = nullptr;
 		if(cargo == "random")
-			commodity = PickCommodity(*source, *result.destination->GetSystem());
+			commodity = PickCommodity(*sourceSystem, *result.destination->GetSystem());
```
Comment (Hurleveur) [inline source/Mission.cpp:1199]: I am not sure why this was changed?
Do note that I am all for just upgrading things in general but I think that this change is unrelated to this PR and not needed. If I remember correclty `source` is used in so many files that changing it only in a few would not make sense, and it is unrelated to this PR.

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -2698,11 +2771,131 @@ void PlayerInfo::CreateMissions()
...
+
+				for(const System* waypoint : mission.Waypoints())
+					destinations.insert(waypoint);
+			}
+		}
+	}
+	availableJobs.sort([&](const Mission &lhs, const Mission &rhs)
 	{
-		return lhs.Name() < rhs.Name();
+		// First, separate rush orders if wanted
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:2806]: I'm not sure what separate rush orders mean, this comment could be more helpful
Rush refers to the missions with a deadline I guess? So in fact this should be named sortMissionByDeadlines?

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -2698,11 +2771,131 @@ void PlayerInfo::CreateMissions()
...
+		if(sortSeparateRush)
+		{
+			if(!lhs.Deadline() && rhs.Deadline())
+				return availableSortAsc; //availableSortAsc instead of true, to counter the actual reversal below
+			if(lhs.Deadline() && !rhs.Deadline())
+				return !availableSortAsc;
+		}
+		// Then, separate greyed-out jobs you can't accept
+		if(sortSeparateGray)
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:2815]: sortSeparateGray doesn't seem very clear either, the color we attribute to jobs that can't be taken shouldn't be used to name this
How about sortPossibleMissions?

File: source/PlayerInfo.h [hunk 6]
```diff
@@ -181,6 +181,17 @@ class PlayerInfo {
 	// Get mission information.
 	const std::list<Mission> &Missions() const;
 	const std::list<Mission> &AvailableJobs() const;
+
+	enum SortType {ABC, PAY, SPEED, CONVENIENT};
```
Comment (Hurleveur) [inline source/PlayerInfo.h:185]: if you look at other enums in the code they are not declared alongside the methods; but in a separate section

**Review by Hurleveur — COMMENTED** [review]
I quite like the idea of sorting the missions, the code looks nice. I feel like the naming of the variables could be improved, also that switch case looks weird to me (the {} are not needed but I guess they make it easier to distinguish what is what).
Also there probably are a few more issues with the style but no big deal.

It seems some things are changed that appear to me a bit outside of the scope of the PR, which doesn't bother me personally but will probably be a blocker for this PR (and could make it less easy to see what the PR actually does).

File: source/PlayerInfo.h (hunk 6 above)
Comment (alextd) [reply source/PlayerInfo.h:185]: I had looked at enums in Mission.h `enum Location` and `enum Trigger`

File: source/MissionPanel.cpp (hunk 2 above)
Comment (alextd) [reply source/MissionPanel.cpp:522]: Yea that's a little redundant. 

But the chunk setting hoverSort is still used elsewhere.

File: source/Mission.cpp (hunk 3 above)
Comment (alextd) [reply source/Mission.cpp:1199]: This one was confusing because `Mission::source` exists and is a Planet instead of a System. The method here defines its own `System *source`. I was meddling with that code since I used that variable.

File: source/PlayerInfo.cpp (hunk 4 above)
Comment (alextd) [reply source/PlayerInfo.cpp:2806]: These are called "separate" instead of "sort" for a few reasons:

They are not the primary sorting method - that's alphabetical, or by pay, etc.

They "sort" by a binary value, instead of sorting a range of possible values: it's more separating true from false, than really sorting a list. 

They are also physically separated when displayed, with an 8-pixel gap between separated missions.

It could be named deadline. I guess all the code calls it though and it's just missions that are named "rush delivery"

**Comment by alextd** [comment]
> switch case looks weird to me (the {} are not needed

The {} for each case allows local variables inside that scope, because otherwise they would be in the `switch` scope, and not initialized in other cases. Kinda dumb, since they aren't *used* in the other cases... but an error nonetheless.

File: source/PlayerInfo.h (hunk 6 above)
Comment (Hurleveur) [reply source/PlayerInfo.h:185]: You're right it, it's not always at the front.

File: source/MissionPanel.cpp [hunk 7]
```diff
@@ -649,13 +687,42 @@ Point MissionPanel::DrawPanel(Point pos, const string &label, int entries) const
 		edgePos.Y() -= dy;
 	}
 
+
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:690]:
```suggestion
```
This extra whitespace doesn't feel like it's needed to me.

File: source/MissionPanel.cpp [hunk 8]
```diff
@@ -649,13 +687,42 @@ Point MissionPanel::DrawPanel(Point pos, const string &label, int entries) const
...
 	const Font &font = FontSet::Get(14);
 	pos += Point(10., 10. + (20. - font.Height()) * .5);
-	font.Draw(label, pos, selected);
+
+	// Panel sorting
+	const Color &text= *GameData::Colors().Get("medium");
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:695]: It's a small thing, but we could be initiating this on line 663, and then get the separatorLine based on it (also because its not a coincidence they both use medium sized colour), also that way all colours are defined at the start of the function.

File: source/MissionPanel.cpp [hunk 9]
```diff
@@ -649,13 +687,42 @@ Point MissionPanel::DrawPanel(Point pos, const string &label, int entries) const
...
+			FillShader::Fill(pos + Point(SIDE_WIDTH - 45., 8.), Point(22., 16.), highlight);
+
+		font.Draw({"? ", {0, Alignment::RIGHT}}, pos + Point(SIDE_WIDTH - 77., 0.), text);
+		SpriteShader::Draw(checkbox[player.ShouldSortSeparatePossible()], pos + Point(SIDE_WIDTH - 70., 8.));
+
+		SpriteShader::Draw(fast, pos + Point(SIDE_WIDTH - 113., 8.));
+		SpriteShader::Draw(checkbox[player.ShouldSortSeparateDeadline()], pos + Point(SIDE_WIDTH - 100., 8.));
+	}
+
+
```
Comment (Hurleveur) [inline source/MissionPanel.cpp:718]:
```suggestion
```
Ik I'm being picky here but its probably because I can't find anything else that can be improved ^^

**Review by Hurleveur — APPROVED** [review]
I just tried it, wow is all that comes to my mind.
This is great! It fits with the style of ES, its smooth, easy to understand... I love it.

Only thing that could maybe be improved is the convenient sorting type image (smth that shows that its on the way; we're already going there) and the `?` used for jobs that can be accepted, I feel like it could be clearer, but I don't have a concrete idea on what it should be, and it's already good enough for me this way. (maybe a check, because those are the missions one can accept?)

I left some suggestions but I'm mostly being picky, the code is great.

**Comment by alextd** [comment]
Yup, the iconography here is just what I thought of. 

* **?** meaning "Can I do this?"
* The 'speedy planet' icon meaning "it's fast to get there" - so it's close-by.
* The 'target reticle' icon meaning "This destination is already a target"

I welcome ideas for icons that would fit in a 22x16 pixel icon that would work. I do imagine a mini-screenshot of the galaxy map, highlighting your path or highlighting the closest planets - but that won't fit here at all.

Also, having played with this a bit, I don't end up using the 'speedy' sort. I do sometimes use alphabetical, surprisingly - there are some jobs I'm looking to do by name, e.g. "Mysterious dropoff"

**Comment by petervdmeer** [comment]
Some review-feedback based on the screenshots (I didn't look at the code yet):
- I like the option to filter out rush-jobs.
- The in-game currency is Credits, not Dollars, so using a Dollar sign seems a bit "out of lore".
- The UI would be less cluttered if the icons themselves function as a checkboxes/toggle-switches (by having a "selected" and a "non selected" version of the filter-icons, and by toggling by clicking on the icons).
- Showing jobs for destinations already in the players current jobs seems to be a "toggle" option, not a sort option. (And is indeed a nice QoL feature.)
- The filters are nice, but the UI should not reveal information not known to the player. The game is primarily a story-telling engine, and revealing information at the wrong moment can break immersion. So sorting by distance should therefore only be done using distances known to the player, and revealing payment information is an issue since we have jobs where the actual payment doesn't match the payment as presented in the mission.
- If I click on a filter because I decided that I don't want to see something, then I expect the specific filtered-out items to disappear from the list (not move to the end of the list).
- Moving unavailable jobs to the end of the list also seems nice, but maybe we better just always do that.

**Comment by Terin** [comment]
> ...and revealing payment information is an issue since we have jobs where the actual payment doesn't match the payment as presented in the mission.

We'll definitely need a `displayed payment` attribute for jobs like A wolf, a goat, and a cabbage, or the spacediving disaster missions, so they can fit naturally in that payment sorting, yeah.

**Comment by MasterOfGrey** [comment]
> * If I click on a filter because I decided that I don't want to see something, then I expect the specific filtered-out items to disappear from the list (not move to the end of the list).
> * Moving unavailable jobs to the end of the list also seems nice, but maybe we better just always do that.

I'm going to disagree on the first point, not because it's wrong from an experience with other software in life settings perspective, but because it is hazardous to game experience. It'd be easy to filter for just the things you're prioritising, and then miss opportunities that you would never take *over* something you're filtering for, but might take with a shortage of options.
From a gameplay angle that would make completely hiding them a detriment.

I will back though that unavailable jobs should maybe just be always moved to the end of the list, that seems like a good QoL thing that will also encourage people to consider what goes into that unavailability as a matter of normal thinking.

**Comment by petervdmeer** [comment]
> I'm going to disagree on the first point, not because it's wrong from an experience with other software in life settings perspective, but because it is hazardous to game experience. It'd be easy to filter for just the things you're prioritising, and then miss opportunities that you would never take _over_ something you're filtering for, but might take with a shortage of options.
> From a gameplay angle that would make completely hiding them a detriment.

I understand your reasoning, and I would expect filters to be "switched off" by default every time the map is opened (the filters are not persistently saved). So every time I open the map I always start with seeing all jobs. But when I do see a list of 60 jobs, and then decide that I don't want to see some of the jobs, then they disappear because I explicitly choose to do so at that moment.

**Comment by samrocketman** [comment]
@alextd are you around to do some additional changes to this?

- Adding tooltips to the icons would be good because they're not necessarily self explanatory.
- The target icon could be a waypoint icon; reference https://discord.com/channels/251118043411775489/431496424992014347/1028145742872846437

---

Re: peter and grey

With regard to filters removing jobs from the list from a play-ability perspective it didn't bother me that jobs moved to the end.

**Review by samrocketman — COMMENTED** [review]
I have done a code style review.  Rather than submitting a bunch of comments in this pull request I have surfaced all code style requests into an easy to merge PR https://github.com/alextd/endless-sky/pull/3 for @alextd 

If you merge this all code style issues will be addressed.

**Comment by alextd** [comment]
Hello this is now updated.

**Comment by alextd** [comment]
Hello this is now updated.

### New UI icons and tooltips:
image

image
image
image


### Addressed issues
Added an "apparent payment" attribute to missions for missions whose actual payment doesn't match the description - e.g. the wolf/cabbage mission. I searched mission files for 'x credit' to find any exact credit references and found Skydiving missions as well. Another mission qualified for "apparent payment" but isn't on the Job Board so it isn't sorted, but why not, it got the attribute as well.

Separating deadline/possible jobs instead of hiding is probably best, as it is kept persistent across systems. Because when you're on a long chain of jobs, you will continue to avoid taking most rush missions - but also there could be a rush mission luckily on the way, so it'll still be there, not completely hidden.


### Lingering not-so-important things:
> The in-game currency is Credits, not Dollars, so using a Dollar sign seems a bit "out of lore".
If there is a credit icon I could certainly use it!

> Showing jobs for destinations already in the players current jobs seems to be a "toggle" option, not a sort option.
The idea was to include planets along the route though I didn't end up programming that entirely. It does sort matching planets > matching system (ie. a different planet in system you're already going to) > other systems though.

> So sorting by distance should therefore only be done using distances known to the player
This is *not* done - with a rationale that the payment for missions is already adjusted for distance --- so you can already sort of glean distance information from jobs. I'd imagine the person sending you on the job tells you how many jumps it'll be, but I also don't want to think too much about the realism of "not having a map of the galaxy"

**Comment by MasterOfGrey** [comment]
> I would expect filters to be "switched off" by default every time the map is opened (the filters are not persistently saved). So every time I open the map I always start with seeing all jobs.

This would frustrate me to no end. In fact I would go so far as to say this would make the feature _nearly_ useless as something intended to make your life easier/more convenient.

**Comment by samrocketman** [comment]
I am playing with this now and it is awesome.
